metadata-ingestion/src/datahub/ingestion/source/vertexai.py

```diff
@@ -0,0 +1,795 @@
+        self.datasets: Optional[dict] = None
+
+    def get_report(self) -> SourceReport:
+        return self.report
+
+    def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Main Function to fetch and yields work units for various VertexAI resources.
+        - Models and Model Versions from the Model Registry
+        - Training Jobs
+        """
+
+        # Ingest Project
+        yield from self._get_project_workunit()
+        # Fetch and Ingest Models, Model Versions a from Model Registry
+        yield from self._get_ml_model_workunits()
+        # Fetch and Ingest Training Jobs
+        yield from self._get_training_jobs_workunit()
+        # TODO Fetch Experiments and Experiment Runs
+
+    def _get_project_workunit(self) -> Iterable[MetadataWorkUnit]:
+        container_key = ProjectIdKey(
+            project_id=self.config.project_id, platform=self.platform
+        )
+
+        yield from datahub.emitter.mcp_builder.gen_containers(
+            container_key=container_key,
+            name=self.config.project_id,
+            sub_types=["Project"],
+        )
+
+    def _validate_training_job(self, model: Model) -> bool:
+        """
+        Validate Model Has Valid Training Job
+        """
+        job = model.training_job
+        if not job:
+            return False
+
+        try:
+            # when model has ref to training job, but field is sometimes not accessible and RunTImeError thrown when accessed
+            # if RunTimeError is not thrown, it is valid and proceed
+            name = job.name
+            logger.debug(
+                (
+                    f"can fetch training job name: {name} for model: (name:{model.display_name} id:{model.name})"
+                )
+            )
+            return True
+        except RuntimeError:
+            logger.debug(
+                f"cannot fetch training job name, not valid for model (name:{model.display_name} id:{model.name})"
+            )
+
+        return False
+
+    def _get_ml_model_workunits(self) -> Iterable[MetadataWorkUnit]:
+        """
+        Fetch List of Models in Model Registry and generate a corresponding work unit.
+        """
+        registered_models = self.client.Model.list()
+        for model in registered_models:
+            # create work unit for Model Group (= Model in VertexAI)
+            yield from self._get_ml_group_workunit(model)
+            model_versions = model.versioning_registry.list_versions()
+            for model_version in model_versions:
+                # create work unit for Training Job (if Model has reference to Training Job)
+                if self._validate_training_job(model):
+                    logger.info(
+                        f"Ingesting a training job for a model: {model_version.model_display_name}"
+                    )
+                    if model.training_job:
+                        yield from self._get_data_process_properties_workunit(
```

Comment:
in general, a lot of your methods are named "get_xyz" but actually produce workunits. some of these methods do actually also fetch data, but not all of them. this makes the code more confusing / difficult to read than it needs to be. but this should get cleaned up through the source

on this particular method, I'd suggest this tweak (also plural workunits, since it returns an iterable)
```suggestion
                        yield from self._gen_data_process_properties_workunits(
```

Reply:
fixed

Reply:
i did naming change throughout,  also did some refactoring as well.